### PR TITLE
[PM-3125] Initial setup on smaller mobile-only sdk

### PR DESCRIPTION
**Pull request by dani-garcia**
## Type of change
```
- [ ] Bug fix
- [x] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective
Create a smaller client surface for the new mobile apps, by hiding all the secrets manager stuff under a `secrets` feature flag, and making a new encryption-focused API hidden under a `mobile` feature flag.

As an initial pass, I've implemented the KDF functions needed for login and the basics needed to initialise the cryptography parts of the client.

Sidenote: We really need to split the bitwarden_json::Command struct into multiple files 😨

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -16,7 +16,7 @@
   "scripts": {
     "lint": "prettier --check .",
     "prettier": "prettier --write .",
-    "schemas": "rimraf ./support/schemas && cargo run --bin sdk-schemas --features internal && ts-node ./support/scripts/schemas.ts",
+    "schemas": "rimraf ./support/schemas && cargo run --bin sdk-schemas --features internal,secrets,mobile && ts-node ./support/scripts/schemas.ts",
```
Comment (Hinton) [inline package.json:19]: Should we use `--all-features` here?

File: crates/bws/Cargo.toml [hunk 2]
```diff
@@ -18,7 +18,7 @@ keywords = ["bitwarden", "secrets-manager", "cli"]
 clap = { version = "4.3.0", features = ["derive", "env"] }
 tokio = { version = "1.28.2", features = ["rt-multi-thread", "macros"] }
 log = "0.4.18"
-bitwarden = { path = "../bitwarden", version = "0.2.1" }
+bitwarden = { path = "../bitwarden", version = "0.2.1", features = ["secrets"] }
```
Comment (Hinton) [inline crates/bws/Cargo.toml:21]: Napi also needs secrets right?

File: crates/bitwarden/src/mobile/client_kdf.rs [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+}
+
+impl<'a> ClientKdf<'a> {
+    pub async fn set_kdf_params(&mut self, req: KdfParamRequest) -> Result<()> {
+        set_kdf_params(self.client, req).await
+    }
+
+    pub async fn get_user_password_hash(&mut self, req: PasswordHashRequest) -> Result<String> {
+        get_user_password_hash(self.client, req).await
+    }
```
Comment (Hinton) [inline crates/bitwarden/src/mobile/client_kdf.rs:16]: Do we need to persist the kdf params or could they be an input to `get_user_password_hash`?

And should we rename it to `hash_password`.

File: crates/bitwarden-json/src/command.rs [hunk 4]
```diff
@@ -1,21 +1,29 @@
...
-    secrets_manager::{
-        projects::{
-            ProjectCreateRequest, ProjectGetRequest, ProjectPutRequest, ProjectsDeleteRequest,
-            ProjectsListRequest,
-        },
-        secrets::{
-            SecretCreateRequest, SecretGetRequest, SecretIdentifiersRequest, SecretPutRequest,
-            SecretsDeleteRequest,
-        },
+use bitwarden::auth::request::AccessTokenLoginRequest;
```
Comment (Hinton) [inline crates/bitwarden-json/src/command.rs:1]: Access token is a secrets manager concept.

File: crates/bitwarden-json/src/command.rs [hunk 5]
```diff
@@ -166,3 +181,40 @@ pub enum ProjectsCommand {
...
+#[serde(rename_all = "camelCase", deny_unknown_fields)]
+pub enum MobileCryptoCommand {
+    /// > Requires having previously set the KDF parameters
+    /// Decrypts the users keys and initializes the user crypto, allowing for the encryption/decryption of the users items
+    ///
+    InitUserCrypto(InitUserCryptoRequest),
+    /// > Requires having previously initialized the user crypto
+    /// Decrypts the users organization keys, allowing for the encryption/decryption of the items in the users organizations
+    ///
+    InitOrgCrypto(InitOrgCryptoRequest),
```
Comment (Hinton) [inline crates/bitwarden-json/src/command.rs:219]: Should this be a single method? I don't think there are any valid scenarios where we only have one of these keys.

File: crates/bitwarden/Cargo.toml [hunk 6]
```diff
@@ -13,7 +13,11 @@ edition = "2021"
 rust-version = "1.57"
 
 [features]
+default = ["secrets"]
```
Comment (Hinton) [inline crates/bitwarden/Cargo.toml:16]: Is this for backwards compatibility? We should add a deprecation note and update our docs to suggest turning this on.

**Review by Hinton — COMMENTED** [review]
Looks like a great start, some thoughts if we can make the hash stateless, and combine the init method.

File: crates/bitwarden/src/mobile/client_kdf.rs (hunk 3 above)
Comment (dani-garcia) [reply crates/bitwarden/src/mobile/client_kdf.rs:16]: We use the KDF params both to create the master password hash for login, and to decrypt the symmetric encryption key in `InitUserCrypto`. 

The reason for the separation in two steps is for a vault unlock scenario, where you already have the vault and encryption keys locally. In those cases, you'd only need to call `set_kdf_params` and then initialize the crypto. But we could just require the KDF parameters as input for the crypto initialization as well.

Agreed on the rename.

File: crates/bitwarden-json/src/command.rs (hunk 5 above)
Comment (dani-garcia) [reply crates/bitwarden-json/src/command.rs:219]: I made those to be one to one with how we handle things internally, but you are right that from a consumer pov the separation is not needed.

File: crates/bitwarden-json/src/command.rs (hunk 4 above)
Comment (dani-garcia) [reply crates/bitwarden-json/src/command.rs]: It's hidden under the feature flag now.

File: crates/bitwarden/Cargo.toml (hunk 6 above)
Comment (dani-garcia) [reply crates/bitwarden/Cargo.toml:16]: Yeah, it's for backwards compatibility and should be removed eventually, I think. I've updated the bitwarden crate README and the docs to include the feature, we should add a deprecation entry to the changelog once that is merged as well.

File: crates/bitwarden/src/mobile/client_kdf.rs (hunk 3 above)
Comment (Hinton) [reply crates/bitwarden/src/mobile/client_kdf.rs]: Hmm that's a valid point.

It does mean on say account switching they would need to call both `set_kdf_params` and `InitCrypto`, which might be a bit confusing.

File: crates/bitwarden/Cargo.toml (hunk 6 above)
Comment (dani-garcia) [reply crates/bitwarden/Cargo.toml:16]: Added the notice to the changelog

File: crates/bitwarden/src/mobile/client_kdf.rs (hunk 3 above)
Comment (dani-garcia) [reply crates/bitwarden/src/mobile/client_kdf.rs]: I've removed the set_kdf_params now, both options seem to have some pros and cons, but at least this way we reduce the API surface a bit

**Review by Hinton — COMMENTED** [review]
Looks good, can we hide the `InitCrypto` command behind the internal flag? It's not yet a stable interface.

**Comment by dani-garcia** [comment]
Do we want to make it so both `mobile` and  `internal` need to be enabled for it to be accessible?

**Comment by Hinton** [comment]
Yes.

File: crates/bitwarden/src/client/auth_settings.rs [hunk 7]
```diff
@@ -31,7 +32,13 @@ pub enum Kdf {
 }
 
 impl AuthSettings {
+    #[cfg(feature = "internal")]
     pub fn new(response: PreloginResponseModel, email: String) -> Self {
+        use crate::util::{
+            default_argon2_iterations, default_argon2_memory, default_argon2_parallelism,
+            default_pbkdf2_iterations,
+        };
```
Comment (Hinton) [inline crates/bitwarden/src/client/auth_settings.rs:40]: What is the benefit of moving use into functions? Just take advantage of the existing `feature` check?

**Review by Hinton — APPROVED** [review]
Looks good!

File: crates/bitwarden/src/client/auth_settings.rs (hunk 7 above)
Comment (dani-garcia) [reply crates/bitwarden/src/client/auth_settings.rs:40]: Yeah, pretty much. Otherwise it's another cfg that we need to add.

**Pull request merged**